android/src/main/java/com/vydia/RNUploader/UploadWorker.kt

```diff
@@ -0,0 +1,267 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.sync.Semaphore
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.util.concurrent.TimeUnit
+
+// All workers will start `doWork` immediately but only 1 request is active at a time.
+const val MAX_CONCURRENCY = 1
+
+// Throttling interval of progress reports
+const val PROGRESS_INTERVAL = 500 // milliseconds
+
+// Retry delay
+val RETRY_DELAY = TimeUnit.SECONDS.toMillis(10L)
+
+// Max total time for a single request to complete
+// This is 24hrs so plenty of time for large uploads
+// Worst case is the time maxes out and the upload gets restarted.
+val REQUEST_TIMEOUT_MILLIS = TimeUnit.HOURS.toMillis(24L)
+
+// Control max concurrent requests using semaphore to instead of using
+// `maxConnectionsCount` in HttpClient as the latter introduces a delay between requests
+val semaphore = Semaphore(MAX_CONCURRENCY)
+private val client = HttpClient(CIO) {
+  install(HttpTimeout) {
+    requestTimeoutMillis = REQUEST_TIMEOUT_MILLIS
+  }
+}
+
+private enum class Connectivity { NoWifi, NoInternet, Ok }
+
+class UploadWorker(private val context: Context, params: WorkerParameters) :
+  CoroutineWorker(context, params) {
+
+  enum class Input { Params }
+
+  private lateinit var upload: Upload
+  private var retries = 0
+  private var connectivity = Connectivity.Ok
+
+  override suspend fun doWork(): Result = withContext(Dispatchers.IO) {
+    // Retrieve the upload. If this throws errors, error reporting won't work.
+    // However, the only way it has errors is the implementation is incorrect,
+    // which can be caught in development
+    val paramsJson = inputData.getString(Input.Params.name) ?: throw Throwable("No Params")
+    upload = Gson().fromJson(paramsJson, Upload::class.java)
+
+    // initialization, errors thrown here won't be retried
+    try {
+      // `setForeground` is recommended for long-running workers.
+      // Foreground mode helps prioritize the worker, reducing the risk
+      // of it being killed during low memory or Doze/App Standby situations.
+      // ⚠️ This should be called in the foreground
+      setForeground(getForegroundInfo())
+    } catch (error: Throwable) {
+      if (!checkAndHandleCancellation()) handleError(error)
+      throw error
+    }
+
+
+    // Complex work, errors thrown below here trigger retry.
+    // We don't let WorkManager manage retries and network constraints as it's very buggy.
+    // i.e. we'd occasionally get BackgroundServiceStartNotAllowedException,
+    // or ForegroundServiceStartNotAllowedException, or "isStopped" gets set to "true"
+    // for no reason
+    var isRetried = false
+    while (true) {
+      try {
+        // Delay if it's an actual retry.
+        // Should be within the try block to account for worker cancellation,
+        // which resumes immediately and throws CancellationException.
+        if (isRetried) delay(RETRY_DELAY)
```

Comment:
Any reason not to implement an exponential backoff here? If there is any persistent source of errors (e.g. the backend could be on the fritz briefly) it seems like we'll quickly fail.

Reply:
good question. One of the reasons is we also use this for invalid connections. 10s seems to be a good maximum wait time when you switch from an invalid connection to a valid connection before the upload starts. The other being we don't consider 400s and 500s as exceptions but success. This is consistent with iOS behavior. If there's an error, a notification will be sent at the end of all uploads so they can manually perform the retry. Since this is a rare case, I'm guessing this is an ok behavior. If the server ends up getting flaky too often we can consider adding exponential backoff just for the 500s errors.

Reply:
- [ ] add this comment to the code